### DISCUSSION TITLE
FAILURE: Build failed with an exception.

### DISCUSSION THREAD

**Discussion by afterlife77** [beeware/briefcase#606] [Q&A]
`sudo briefcase create` &` sudo briefcase build` complete successfully. Problem starts when I try to `build` for android . . .
`sudo briefcase create android` works but `sudo briefcase build android` throws exception;

[helloworld] Building Android APK...

```
> Task :app:externalNativeBuildDebug
Build native-lib_armeabi-v7a
ninja: no work to do.
Build native-lib_arm64-v8a
ninja: no work to do.
Build native-lib_x86
ninja: no work to do.
Build native-lib_x86_64
ninja: no work to do.

> Task :app:stripDebugDebugSymbols FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:stripDebugDebugSymbols'.
> No toolchains found in the NDK toolchains folder for ABI with prefix: arm-linux-androideabi

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
27 actionable tasks: 2 executed, 25 up-to-date

```

When I run `build` without `sudo`;

```
[helloworld] Building Android APK...

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type ScriptPluginFactory using BuildScopeServices.createScriptPluginFactory().
> Could not create service of type FileHasher using BuildSessionScopeServices.createFileSnapshotter().

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s

Error while building project.
```

Same with or without venv . . .

Also I messed a little with `docker` but I don't really know how to use it as when I try to run `python` or `pip` it says; `command not found`,and `apt` has none of those packages. Maybe that is the underlying issue.

Anyone?

**Comment by freakboy3742** [comment]
First off - you shouldn't need to use `sudo` at all. If you do, then there's a much bigger problem going on - which is may be the cause of the later errors you've provided. Briefcase should *always* be run as a user command, and should be able to succeed without any superuser permissions. If briefcase doesn't succeed, *that* is the first problem to fix. "Just use sudo" isn't a fix.

So - my first suggestion - start from scratch. Delete the tutorial folder and the `~/.briefcase` folder, then start from scratch, *without* using sudo.

If that doesn't resolve things, and you get the same error - the error you're reporting means that NDK hasn't been downloaded. This should be downloaded automatically the first time an Android project is built; if it's not being found and it's not being downloaded, then there is some form of system corruption getting in the way.

The most likely possibility is that a pre-existing Android Studio install is getting in the way. Have you previously installed Android Studio? Are there any environment variables set (like `ANDROID_HOME` or `ANDROID_SDK_ROOT`)? If you remove these variables from the shell environment and re-run `briefcase build android` - does the build work?

If you haven't installed Android Studio, and a non-sudo build doesn't work, then we'll need to look elsewhere...

**Reply by afterlife77** [reply]
> First off - you shouldn't need to use `sudo` at all. If you do, then there's a much bigger problem going on - which is may be the cause of the later errors you've provided. Briefcase should _always_ be run as a user command, and should be able to succeed without any superuser permissions. If briefcase doesn't succeed, _that_ is the first problem to fix. "Just use sudo" isn't a fix.
> 
> So - my first suggestion - start from scratch. Delete the tutorial folder and the `~/.briefcase` folder, then start from scratch, _without_ using sudo.
> 
> If that doesn't resolve things, and you get the same error - the error you're reporting means that NDK hasn't been downloaded. This should be downloaded automatically the first time an Android project is built; if it's not being found and it's not being downloaded, then there is some form of system corruption getting in the way.
> 
> The most likely possibility is that a pre-existing Android Studio install is getting in the way. Have you previously installed Android Studio? Are there any environment variables set (like `ANDROID_HOME` or `ANDROID_SDK_ROOT`)? If you remove these variables from the shell environment and re-run `briefcase build android` - does the build work?
> 
> If you haven't installed Android Studio, and a non-sudo build doesn't work, then we'll need to look elsewhere...

I removed `~/.briefcase` and `helloworld` folder. I used `unset ANDROID_HOME` &  `unset ANDROID_SDK_ROOT ` so my `/etc/environment` contains only `PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"`.

`briefcase build android` & `python3 -m briefcase build android` still return the NDK toolchains exception.

`python -m pip install briefcase` returns 

```
Command 'python' not found, did you mean:

  command 'python3' from deb python3
  command 'python' from deb python-is-python3
```




Originally occurred on kali linux, reproduced error on ubuntu 20. 

Edit: I haven't installed `toga` with `pip` as I realized briefcase manages dependencies.

**Reply by freakboy3742** [reply]
If you're getting errors about `python` not working, that tells me you're missing some steps in the tutorial - most notably, you're probably not using a virtual environment.

However, that isn't causing the NDK problem you're reporting - I've now been able to resolve this problem; something has evidently changed in NDK, and the binary targets are now slightly different. If you had an old version of NDK downloaded (as I did), the project works; the current download doesn't work. I'll need to investigate further.

**Reply by freakboy3742** [reply]
See #607 for a formal bug report for this problem.
